### ISSUE TITLE
Documentation for context manager?

### ISSUE THREAD

**Issue by Microserf** [kennethreitz/requests#2580]
The documentation makes only a single reference to the fact that `requests` provides some sort of context manager, and it's in the release notes from version 0.6 in 2011:

> New persistient sessions object and context manager

Would it be possible to get examples of how to implement these context managers? Specifically, I'm looking to replace the following code:

```python
    with contextlib.closing(urllib2.urlopen(self.url)) as dl:
```

I can't find any place that documents how to use an equivalent `requests` context manager, so I'll have to dig through the code to find out. It would be nice to get a bit more visibility into the `requests` API...

**Comment by sigmavirus24** [comment]
So I believe that refers to the fact that you can do:

``` py
with requests.Session() as session:
    response = session.get(self.url)
```

To get the equivalent to that code though, you could do

``` py
import requests

session = requests.Session()

with contextlib.closing(session.get(self.url, stream=True)) as response:
    response.raw.read(...)

# or

with contextlib.closing(requests.get(self.url, stream=True)) as response:
```

I could have sworn there were well documented examples of using Sessions and responses as context managers.

**Comment by Lukasa** [comment]
> I could have sworn there were well documented examples of using Sessions and responses as context managers.

Nothing I can find in the docs, only your `contextlib.closing` example from above. Feels like a good enhancement!

**Comment by Microserf** [comment]
@sigmavirus24 Thanks for the quick response with examples. I'll try these and report back...

**Comment by Microserf** [comment]
Update: I got the second form (the one with `requests.get` instead of `session.get`) to work. Didn't test the first form.

I had to remember to use `response.raw` as the file-like object, not `response`. (This is different from the `urllib2.urlopen()` example I gave above, where `dl.read()` gives the response). This makes sense, when I considered how the code wrapped within the context manager may want to check the response status, encoding types or other metadata, before reading the contents of the response.

Thanks!
